### PR TITLE
Add a REPL widget that supports the SciJava ScriptREPL

### PR DESCRIPTION
**Pull request by kephale**
This PR adds a widget for the SciJava ScriptREPL, allowing users to interactively write SciJava friendly code directly within napari.

When I was exploring the compatibility of various ImageJ/Fiji plugins with napari-imagej, I frequently found myself wanting to poke at variables/state. That led me to writing this REPL widget so we can quickly poke around on the JVM instead of always needing to deal with jpipe/jimport stuff.

napari_imagej_repl_001
Caption: This is a screenshot of napari-imagej with the REPL widget active and some example output from  Jython.

Now updated to have the corrected icon:
image

### CODE REVIEW

**Comment by kephale** [comment]
BTW, one thing I noticed with this is that the `Clojure` language actually crashes this plugin. It looks like one of the Clojure classes from the main clojure jar does not get found. All of the other languages work.

**Comment by gselzer** [comment]
I'm not sure that I'll have the chance to look at the code until Tuesday or so @kephale, but this looks really awesome - thanks for implementing it!

One quick comment - have you considered adding the REPL to the bottom by default? I really like how the napari console sits at the bottom - naively this seems like it should go in the same place.

**Comment by kephale** [comment]
I'm very open to where it goes. I have some concerns about all the consoles/repls getting confused. I was actually a little worried about making this icon so similar to the napari-console icon.

**Comment by gselzer** [comment]
> I have some concerns about all the consoles/repls getting confused. I was actually a little worried about making this icon so similar to the napari-console icon.

Yeah, I thought about that too.

As far as the button goes, maybe we could color it differently, or add the ImageJ Logo somewhere? Another option would be to just make the SciJava Script REPL a separate napari plugin - making it show up in the napari menus - then we don't have to worry about a button.

Then there's the question of differentiating the two consoles - does the SciJava Script REPL have a Java UI that we could launch (i.e. launch a separate window), and if so, would that detract from your probing properties if it was in a different window? That could be one way to do it, and would require less Qt code to maintain.

**Comment by kephale** [comment]
> SciJava Script REPL have a Java UI

Yes, but I was particularly keen on this for the headless situation, so I do hope that the UI is on the Python side.

**Review by gselzer — CHANGES_REQUESTED** [review]
The code itself looks great @kephale, just had one minor suggestion.

I think some documentation, both in the GUI (tooltips) and in the docs would be warranted (maybe this could help in troubleshooting?)

**Comment by kephale** [comment]
@gselzer I emailed you 1 more figure for this addition to the docs: https://github.com/kephale/napari-imagej/blob/4ccab72c0d8e2890e9ed27d8e4b5e1bff3da547e/doc/Configuration.rst?plain=1#L106

I've added tooltips + a section in the docs about the REPL.

File: doc/Configuration.rst [hunk 1]
```diff
@@ -95,6 +95,15 @@ One common use case for this feature is to increase the maximum heap space avail
 
     Specifying 32GB of memory available to ImageJ ecosystem routines in the JVM.
 
+Using the SciJava REPL
+--------------------------------
+
+You can use the SciJava REPL to interactively run SciJava code. This makes it possible to do things like paste existing SciJava scripts into the REPL. More information on scripting in SciJava can be found `here <https://imagej.net/scripting/>`_.
```
Comment (gselzer) [inline doc/Configuration.rst:101]: So I played a little with the Script REPL, and I couldn't get anything beyond `2+4` to work. I was trying to print `Hello World`, but I couldn't even figure out how to do that.

So, as a result, I looked here, and I checked out imagej.net/scripting. As someone who is not intimately familiar with that page, I wasn't super sure where to go. I did find [the ScriptInterpreter page](https://imagej.net/scripting/interpreter) from that page, which *appears to me to be what this is*, but I tried typing `:help` and I didn't get anything.

Then, I typed `help`, and my napari froze. I went back to my terminal, and I saw this:

```bash
(napari-imagej-dev) ➜  napari-imagej git:(scijava-repl) napari
hello
[java.lang.Enum.toString] hello[java.lang.Enum.toString] 
Hello World
[java.lang.Enum.toString] Hello World[java.lang.Enum.toString] 
Hello World
[java.lang.Enum.toString] Hello World[java.lang.Enum.toString] 
Hello World
[java.lang.Enum.toString] Hello World[java.lang.Enum.toString] 
Hello World
[java.lang.Enum.toString] Hello World[java.lang.Enum.toString] 

Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> [java.lang.Enum.toString] 
Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> 
```
So, long story short, it looks like the output is not being captured by the TextEdit, and also I'm wondering if we can link to documentation that is more specific to the SciJava REPL.

To solve the output not being captured by the TextEdit, is there an option in SciJava for this @ctrueden?

File: doc/Configuration.rst [hunk 2]
```diff
@@ -95,6 +95,15 @@ One common use case for this feature is to increase the maximum heap space avail
 
     Specifying 32GB of memory available to ImageJ ecosystem routines in the JVM.
 
+Using the SciJava REPL
+--------------------------------
```
Comment (gselzer) [inline doc/Configuration.rst:99]: I'm wondering if this would be better suited to the new `Usage` page of the ReadTheDocs...

Comment (gselzer) [reply doc/Configuration.rst:99]: Or the `Troubleshooting` page...

File: src/napari_imagej/widgets/menu.py [hunk 3]
```diff
@@ -255,6 +266,45 @@ def disable_popup(self):
...
 
+class REPLButton(IJMenuButton):
+    def __init__(self, viewer: Viewer):
+        super().__init__(viewer)
+        self.viewer = viewer
+
+        self.setEnabled(False)
+
+        icon = QColoredSVGIcon(resource_path("repl"))
+        self.setIcon(icon.colored(theme=viewer.theme))
```
Comment (gselzer) [inline src/napari_imagej/widgets/menu.py:277]: If you could replace this with the way that we set the icon [here](https://github.com/imagej/napari-imagej/blob/75cedb62a87a8d8ec20a1669133c517bf48c0b6e/src/napari_imagej/widgets/menu.py#L125), then the icon will be recolored when the user changes the theme.

File: src/napari_imagej/widgets/menu.py [hunk 4]
```diff
@@ -255,6 +266,45 @@ def disable_popup(self):
...
+
+        self.clicked.connect(self._toggle_repl)
+        self._widget = None
+
+    def _add_repl_to_dock(self):
+        from scyjava import jimport
+
+        ByteArrayOutputStream = jimport("java.io.ByteArrayOutputStream")
+        ScriptREPL = jimport("org.scijava.script.ScriptREPL")
```
Comment (gselzer) [inline src/napari_imagej/widgets/menu.py:286]: Can we put these imports in [`java.py`](https://github.com/imagej/napari-imagej/blob/75cedb62a87a8d8ec20a1669133c517bf48c0b6e/src/napari_imagej/java.py#L196), along with all of the other imports? It's kinda nice to use `jimport` in one place

File: src/napari_imagej/widgets/repl.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        Process the user input and evaluate it using the ScriptREPL.
+
+        Display the results in the output text area.
+        """
+        input_text = self.input_lineedit.text()
+        self.input_lineedit.clear()
+
+        from scyjava import jimport
+
+        ScriptException = jimport("javax.script.ScriptException")
```
Comment (gselzer) [inline src/napari_imagej/widgets/repl.py:65]: It would also be cool if this `jimport` was moved to `java.py`.

File: src/napari_imagej/widgets/menu.py [hunk 6]
```diff
@@ -41,7 +43,12 @@ def __init__(self, viewer: Viewer):
         self.gui_button: GUIButton = GUIButton(viewer)
         self.layout().addWidget(self.gui_button)
 
+        self.repl_button: REPLButton = REPLButton(viewer)
+        self.repl_button.setToolTip("Show/hide the SciJava REPL")
+        self.layout().addWidget(self.repl_button)
+
         self.settings_button: SettingsButton = SettingsButton(viewer)
+        self.settings_button.setToolTip("Show napari-imagej settings")
```
Comment (gselzer) [inline src/napari_imagej/widgets/menu.py:51]: Generally I've been putting the tool tips in the constructors of each button - could we move these there for consistency?

Thanks so much for adding the tooltip to the settings button - can't believe I forgot that!

**Review by gselzer — CHANGES_REQUESTED** [review]
After looking at this again, I'm actually pretty happy with the visual appearance of the REPL - I don't think it's too easy to confuse with the napari console.

One thing that I did think would be cool would be to put a shadow of the napari-imagej logo behind the text edit for a visual reminder of which textedit is which, but I have *no idea* how you'd do that, so I'm just going to leave it here as an idea.

**Comment by ctrueden** [comment]
@gselzer It seems you started reviewing this as I was actively trying to fix the same problems you discovered. This PR needs some additional work to squash before it can be merged. If you want to get looped into fixing it (which would be really helpful!), I can give you a primer on how the SciJava `ScriptREPL` works on Monday.

**Comment by ctrueden** [comment]
I'm still working on this. Need to change `ScriptREPL` on the scijava-common side to accept an output handler that is not an `OutputStream`. My plan is to do a generic `Consumer<String>` and deprecate the existing `OutputStream` signature to use a lambda that just passes it as a `Consumer`.

**Comment by ctrueden** [comment]
Progress: scijava/scijava-common@0f3bf4449bbcb5bcce5ec5362e2d92bae39619b4. Was going to release it, but updating scijava-common to pom-scijava 36.0.0 breaks some tests, so gotta sort that out first. Then we can start using the improved `ScriptREPL` on this PR to finally sort out its input/output handling.

**Comment by ctrueden** [comment]
Progress today: scijava-common 2.95.0 was released. And I worked further on this branch, rebasing it on top of the latest mainline. Some tests pass now, but others fail. I need to scrutinize the design changes a bit more, and probably reshuffle the commit order to minimize the footprint of the commit implementing the changeover to the nij object.

File: src/napari_imagej/widgets/menu.py (hunk 6 above)
Comment (gselzer) [reply src/napari_imagej/widgets/menu.py]: Moved in [6dcacbf](https://github.com/imagej/napari-imagej/pull/265/commits/6dcacbfbe49f0ba256259c1dea77bfe79c7dcf13)

File: src/napari_imagej/widgets/repl.py (hunk 5 above)
Comment (gselzer) [reply src/napari_imagej/widgets/repl.py]: Outdated.

File: src/napari_imagej/widgets/menu.py (hunk 4 above)
Comment (gselzer) [reply src/napari_imagej/widgets/menu.py]: Also outdated.

File: src/napari_imagej/widgets/menu.py (hunk 3 above)
Comment (gselzer) [reply src/napari_imagej/widgets/menu.py]: Fixed with [60f1f77](https://github.com/imagej/napari-imagej/pull/265/commits/60f1f77b89ada2431b3f90e0409e9d3f185f4bfe)

**Comment by gselzer** [comment]
This was a *long* time coming @ctrueden @kephale, however I've just rebased this PR on top of the latest `main` and resolved the WIP, getting tests to pass. I'm embarrased that we never merged this, so I'm making it a priority for the 0.2.0 release 😤.

I've still got an error where the bindings aren't found, I likely just have to find the code adding them on the Java side and copy/call it, but beyond that this REPL performs identically to the one in Fiji 🪄

TODO for merge:
* [ ] Add REPL bindings
* [ ] Consider any easy tests
* [ ] Update documentation
* [ ] (Tangent) update all documentation (probably should make another issue for this)

File: doc/Configuration.rst (hunk 1 above)
Comment (gselzer) [reply doc/Configuration.rst:101]: We now have it where the input and output is being displayed identically to the SciJava ScriptREPL - any further problems, including this one, should be fixed upstream.

File: doc/Configuration.rst (hunk 2 above)
Comment (gselzer) [reply doc/Configuration.rst]: Moved to the `Usage` page.

**Review by gselzer — APPROVED** [review]
